File: packages/go/conformance_test.go
package nebulatoken

// Shared conformance vectors — spec/test-vectors.json (SPECIFICATION.md [N-47]).
//
// The vectors are data published once for all ten implementations; nothing in
// this file re-derives a case by hand. Every section asserts that the number of
// cases it executed equals the published count, so a runner that silently
// iterated zero cases fails instead of passing ([N-48]).

import (
	"encoding/base64"
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
)

type testVectors struct {
	SpecVersion int `json:"spec_version"`
	Counts      struct {
		VerifierHashing int `json:"verifier_hashing"`
		DeviceHashing   int `json:"device_hashing"`
		Parsing         int `json:"parsing"`
	} `json:"counts"`
	Constants       json.RawMessage `json:"constants"`
	VerifierHashing []struct {
		ID       string `json:"id"`
		Pepper   string `json:"pepper"`
		Verifier string `json:"verifier_b64url"`
		Note     string `json:"note"`
		Expected string `json:"expected_hmac_sha256_hex"`
	} `json:"verifier_hashing"`
	DeviceHashing []struct {
		ID       string `json:"id"`
		Pepper   string `json:"pepper"`
		DeviceID string `json:"device_id"`
		Note     string `json:"note"`
		Expected string `json:"expected_hmac_sha256_hex"`
	} `json:"device_hashing"`
	Parsing []struct {
		ID       string `json:"id"`
		Token    string `json:"token"`
		Valid    bool   `json:"valid"`
		Kid      string `json:"kid"`
		Selector string `json:"selector"`
		Note     string `json:"note"`
		Rule     string `json:"rule"`
	} `json:"parsing"`
}

// specDir walks up from this source file to the repository root and returns its
// spec/ directory. Walking beats a hardcoded path: the tests run from the
// package directory locally and from a bind mount in CI, and the vectors are
// never copied into the package — one published copy, ten implementations.
func specDir(t *testing.T) string {
	t.Helper()
	_, thisFile, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatal("cannot locate this source file")
	}
	dir := filepath.Dir(thisFile)
	for {
		candidate := filepath.Join(dir, "spec")
		if _, err := os.Stat(filepath.Join(candidate, "test-vectors.json")); err == nil {
			return candidate
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			t.Fatalf("no spec/test-vectors.json found walking up from %s", filepath.Dir(thisFile))
		}
		dir = parent
	}
}

func loadTestVectors(t *testing.T) testVectors {
	t.Helper()
	raw, err := os.ReadFile(filepath.Join(specDir(t), "test-vectors.json"))
	if err != nil {
		t.Fatalf("cannot read vectors: %v", err)
	}
	var v testVectors
	if err := json.Unmarshal(raw, &v); err != nil {
		t.Fatalf("cannot parse vectors: %v", err)
	}
	return v
}

func TestSpecVersionMatchesVectors(t *testing.T) {
	if got := loadTestVectors(t).SpecVersion; got != SpecVersion {
		t.Fatalf("package implements spec version %d, vectors publish %d", SpecVersion, got)
	}
}

// TestConstants compares every constant the specification publishes against the
// exported value, in both directions: a constant published but not compared is
// a conformance gap, and a comparison of something that is not published is a
// stale test ([N-4], [N-48]).
func TestConstants(t *testing.T) {
	published := map[string]json.RawMessage{}
	if err := json.Unmarshal(loadTestVectors(t).Constants, &published); err != nil {
		t.Fatalf("cannot parse constants section: %v", err)
	}
	if len(published) == 0 {
		t.Fatal("constants section is absent or empty ([N-48])")
	}

	ours := map[string]any{
		"prefix":                       Prefix,
		"selector_bytes":               SelectorBytes,
		"verifier_bytes":               VerifierBytes,
		"selector_chars":               SelectorChars,
		"verifier_chars":               VerifierChars,
		"max_kid_length":               MaxKidLength,
		"max_token_length":             MaxTokenLength,
		"min_pepper_length":            MinPepperLength,
		"default_absolute_ttl_seconds": DefaultAbsoluteTTL,
		"default_idle_ttl_seconds":     DefaultIdleTTL,
		"default_reuse_grace_seconds":  DefaultReuseGrace,
	}

	for name, want := range published {
		got, ok := ours[name]
		if !ok {
			t.Errorf("constant %q is published but never compared ([N-4])", name)
			continue
		}
		encoded, err := json.Marshal(got)
		if err != nil {
			t.Fatalf("%s: %v", name, err)
		}
		if string(encoded) != string(want) {
			t.Errorf("constant %q: package has %s, specification publishes %s", name, encoded, want)
		}
	}
	for name := range ours {
		if _, ok := published[name]; !ok {
			t.Errorf("compared %q, which the specification does not publish", name)
		}
	}
}

func TestVerifierHashingVectors(t *testing.T) {
	v := loadTestVectors(t)
	n := 0
	for _, c := range v.VerifierHashing {
		verifier, err := base64.RawURLEncoding.DecodeString(c.Verifier)
		if err != nil {
			t.Fatalf("%s: unusable vector: %v", c.ID, err)
		}
		if got := HashVerifier(c.Pepper, verifier); got != c.Expected {
			t.Errorf("%s (%s): got %s want %s", c.ID, c.Note, got, c.Expected)
		}
		n++
	}
	if n != v.Counts.VerifierHashing {
		t.Fatalf("executed %d verifier_hashing cases, published count is %d ([N-48])", n, v.Counts.VerifierHashing)
	}
}

func TestDeviceHashingVectors(t *testing.T) {
	v := loadTestVectors(t)
	n := 0
	for _, c := range v.DeviceHashing {
		got, err := HashDeviceID(c.Pepper, c.DeviceID)
		if err != nil {
			t.Fatalf("%s (%s): unexpected error %v", c.ID, c.Note, err)
		}
		if got != c.Expected {
			t.Errorf("%s (%s): got %s want %s", c.ID, c.Note, got, c.Expected)
		}
		n++
	}
	if n != v.Counts.DeviceHashing {
		t.Fatalf("executed %d device_hashing cases, published count is %d ([N-48])", n, v.Counts.DeviceHashing)
	}
}

func TestParsingVectors(t *testing.T) {
	v := loadTestVectors(t)
	n := 0
	for _, c := range v.Parsing {
		parsed := ParseToken(c.Token)
		switch {
		case c.Valid && parsed == nil:
			t.Errorf("%s: expected a valid token (%s)", c.ID, c.Note)
		case c.Valid:
			if parsed.Kid != c.Kid {
				t.Errorf("%s: kid %q, want %q", c.ID, parsed.Kid, c.Kid)
			}
			if parsed.Selector != c.Selector {
				t.Errorf("%s: selector %q, want %q", c.ID, parsed.Selector, c.Selector)
			}
			if len(parsed.Verifier) != VerifierBytes {
				t.Errorf("%s: verifier decoded to %d bytes, want %d", c.ID, len(parsed.Verifier), VerifierBytes)
			}
		case parsed != nil:
			t.Errorf("%s: expected MALFORMED per %s (%s)", c.ID, c.Rule, c.Note)
		}
		n++
	}
	if n != v.Counts.Parsing {
		t.Fatalf("executed %d parsing cases, published count is %d ([N-48])", n, v.Counts.Parsing)
	}
}

// TestParsingIsTotal covers [N-8] for inputs no JSON vector can carry: Go
// strings are arbitrary byte sequences, so they reach the parser holding
// invalid UTF-8, NUL bytes and lone surrogates in WTF-8 form.
func TestParsingIsTotal(t *testing.T) {
	hostile := []string{
		"",
		" ",
		".",
		strings.Repeat(".", 1000),
		"nbl." + strings.Repeat("k", 10000),
		"nbl.k1." + strings.Repeat(" ", 22) + "." + strings.Repeat("A", 43),
		"nbl.k1.\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00." + strings.Repeat("A", 43),
		"\xff\xfe",
		"nbl.\xed\xa0\x80." + strings.Repeat("A", 22) + "." + strings.Repeat("A", 43),
		strings.Repeat("\xff", MaxTokenLength+1),
	}
	for _, in := range hostile {
		// A panic here fails the test by itself; the assertion pins the value.
		if got := ParseToken(in); got != nil {
			t.Errorf("ParseToken(%q) = %+v, want nil", in, got)
		}
	}
}

File: packages/php/tests/ConformanceTest.php
<?php

declare(strict_types=1);

namespace NebulaToken\Tests;

use NebulaToken\Nebula;
use PHPUnit\Framework\TestCase;
use stdClass;

/**
 * Shared conformance vectors — spec/test-vectors.json (SPECIFICATION.md [N-47]).
 *
 * The vectors are read from the repository, never copied into the package, so
 * this suite cannot drift from the other nine implementations.
 */
final class ConformanceTest extends TestCase
{
    /** @return array<string, mixed> */
    private static function vectors(): array
    {
        // Walk up from this file until the repository's spec/ directory shows
        // up: the package must stay relocatable, so no path is hardcoded.
        $dir = __DIR__;
        while (!is_file($dir . '/spec/test-vectors.json')) {
            $parent = dirname($dir);
            self::assertNotSame($parent, $dir, 'could not locate spec/test-vectors.json above ' . __DIR__);
            $dir = $parent;
        }
        $raw = file_get_contents($dir . '/spec/test-vectors.json');
        self::assertNotFalse($raw);

        return json_decode($raw, true, 512, JSON_THROW_ON_ERROR);
    }

    /**
     * php:S3415 reads the arguments as swapped, because the second one is a
     * class constant and its heuristic assumes a constant is always the
     * *expected* value. Here it is the opposite way round, deliberately and
     * throughout this file: the published vector is the expected value — it is
     * the normative artefact every port is measured against — and the library
     * constant is the actual, because the library is what is under test. Swapping
     * them would invert what a failure message means, and would break step with
     * the Java and Ruby conformance suites, which assert this in the same order.
     */
    public function testSpecVersionMatchesThePublishedVectors(): void
    {
        self::assertSame(self::vectors()['spec_version'], Nebula::SPEC_VERSION); // NOSONAR(php:S3415)
    }

    public function testConstantsMatchSpec(): void
    {
        $published = self::vectors()['constants'];
        $compared = [
            'prefix' => Nebula::PREFIX,
            'selector_bytes' => Nebula::SELECTOR_BYTES,
            'verifier_bytes' => Nebula::VERIFIER_BYTES,
            'selector_chars' => Nebula::SELECTOR_CHARS,
            'verifier_chars' => Nebula::VERIFIER_CHARS,
            'max_kid_length' => Nebula::MAX_KID_LENGTH,
            'max_token_length' => Nebula::MAX_TOKEN_LENGTH,
            'min_pepper_length' => Nebula::MIN_PEPPER_LENGTH,
            'default_absolute_ttl_seconds' => Nebula::DEFAULT_ABSOLUTE_TTL,
            'default_idle_ttl_seconds' => Nebula::DEFAULT_IDLE_TTL,
            'default_reuse_grace_seconds' => Nebula::DEFAULT_REUSE_GRACE,
        ];

        // [N-48]: every constant the spec publishes is compared, not only the
        // ones this file happened to remember.
        self::assertSame(
            [],
            array_keys(array_diff_key($published, $compared)),
            'a published constant was never asserted',
        );
        foreach ($compared as $name => $value) {
            self::assertSame($published[$name], $value, $name);
        }
    }

    public function testVerifierHashingVectors(): void
    {
        $vectors = self::vectors();
        self::assertNotEmpty($vectors['verifier_hashing'], 'section absent or empty ([N-48])');

        $n = 0;
        foreach ($vectors['verifier_hashing'] as $v) {
            $verifier = Nebula::b64urlDecode($v['verifier_b64url']);
            self::assertNotNull($verifier, $v['id']);
            self::assertSame(
                $v['expected_hmac_sha256_hex'],
                Nebula::hashVerifier($v['pepper'], $verifier),
                $v['id'] . ': ' . $v['note'],
            );
            $n++;
        }
        self::assertSame($vectors['counts']['verifier_hashing'], $n, 'executed count must equal published count ([N-48])');
    }

    public function testDeviceHashingVectors(): void
    {
        $vectors = self::vectors();
        self::assertNotEmpty($vectors['device_hashing'], 'section absent or empty ([N-48])');

        $n = 0;
        foreach ($vectors['device_hashing'] as $v) {
            self::assertSame(
                $v['expected_hmac_sha256_hex'],
                Nebula::hashDeviceId($v['pepper'], $v['device_id']),
                $v['id'] . ': ' . $v['note'],
            );
            if (isset($v['device_id_bytes'])) {
                // [N-11] keys the HMAC on the UTF-8 encoding of the identifier,
                // not on however the runtime happens to hold it. A PHP string
                // *is* a byte string, so the decoded bytes go straight in with
                // nothing converting them; a runner whose strings carry an
                // encoding separate from their bytes has more to do. Either way
                // the case's one expected hash must come out, which is the
                // portable statement of the rule — and the assertion that a
                // runtime cannot decide a device identifier on anything but its
                // bytes.
                $fromBytes = hex2bin($v['device_id_bytes']);
                self::assertNotFalse($fromBytes, $v['id'] . ': device_id_bytes must be hex');
                self::assertSame(
                    $v['device_id'],
                    $fromBytes,
                    $v['id'] . ': device_id_bytes must be the UTF-8 encoding of device_id',
                );
                self::assertSame(
                    $v['expected_hmac_sha256_hex'],
                    Nebula::hashDeviceId($v['pepper'], $fromBytes),
                    $v['id'] . ' from bytes',
                );
            }
            $n++;
        }
        self::assertSame($vectors['counts']['device_hashing'], $n, 'executed count must equal published count ([N-48])');
    }

    public function testParsingVectors(): void
    {
        $vectors = self::vectors();
        self::assertNotEmpty($vectors['parsing'], 'section absent or empty ([N-48])');

        $n = 0;
        foreach ($vectors['parsing'] as $v) {
            $parsed = Nebula::parseToken($v['token']);
            if ($v['valid']) {
                self::assertNotNull($parsed, $v['id'] . ' should parse: ' . $v['note']);
                self::assertSame($v['kid'], $parsed->kid, $v['id']);
                self::assertSame($v['selector'], $parsed->selector, $v['id']);
                self::assertSame(Nebula::VERIFIER_BYTES, strlen($parsed->verifier), $v['id']);
            } else {
                self::assertNull($parsed, $v['id'] . ' should be MALFORMED (' . $v['rule'] . '): ' . $v['note']);
            }
            $n++;
        }
        self::assertSame($vectors['counts']['parsing'], $n, 'executed count must equal published count ([N-48])');
    }

    public function testParsingIsTotal(): void
    {
        // [N-8]: no input may raise. PHP's string type is a byte string, so
        // invalid UTF-8 reaches the parser routinely; `mixed` covers the
        // non-string values a loosely typed caller can hand over.
        $hostile = [
            null, 42, 4.5, true, false, [], new stdClass(),
            '', ' ', '.', str_repeat('.', 1000),
            'nbl.' . str_repeat('k', 10_000),
            'nbl.k1.' . str_repeat(' ', 22) . '.' . str_repeat('A', 43),
            "nbl.k1.\xED\xA0\x80\xED\xA0\x80\xED\xA0\x80\xED\xA0\x80\xED\xA0\x80\xED\xA0\x80\xED\xA0\x80.\xff",
            "\xff\xfe",
            "nbl.k1.AAECAwQFBgcICQoLDA0ODw.AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8\0",
        ];
        foreach ($hostile as $input) {
            self::assertNull(Nebula::parseToken($input), 'hostile input must parse to null, not raise');
        }
    }
}
